ec/src/hashing/curve_maps/swu/mod.rs

```diff
@@ -0,0 +1,181 @@
+use core::marker::PhantomData;
+
+use ark_ff::bytes::{ToBytes};
+
+use crate::models::SWModelParameters;
+use ark_ff::{vec::Vec, Field, One, SquareRootField, Zero};
+use ark_std::string::ToString;
+
+use crate::{
+    hashing::{map_to_curve_hasher::MapToCurve, HashToCurveError},
+    models::short_weierstrass_jacobian::GroupAffine,
+    AffineCurve,
+};
+
+/// Trait defining the necessary parameters for the SWU hash-to-curve method
+/// for the curves of Weierstrass form of:
+/// y^2 = x^3 + a*x + b where ab != 0. From [WB2019]
+///
+/// - [WB19] Wahby, R. S., & Boneh, D. (2019). Fast and simple constant-time
+///   hashing to the bls12-381 elliptic curve. IACR Transactions on
+///   Cryptographic Hardware and Embedded Systems, nil(nil), 154–179.
+///   http://dx.doi.org/10.46586/tches.v2019.i4.154-179
+pub trait SWUParams: SWModelParameters {
+    // we need an element of the base field which is not a square root see [1] Sect.
+    // 4. it is also convenient to have $g(b/xi * a)$ to be square. In general
+    // we use a xi with low absolute value coefficients when they are
+    // represented as element of ZZ.
+    const XI: Self::BaseField; // a nonsquare in Fq
+    const ZETA: Self::BaseField; // arbitatry root of unity
+    const XI_ON_ZETA_SQRT: Self::BaseField; // square root of THETA
```

Comment:
Q: Is the comment about `THETA` correct?

```suggestion
    const XI: Self::BaseField;
    /// An arbitrary root of unity in the base field.
    const ZETA: Self::BaseField;
    /// Square root of `THETA = Self::XI/Self::ZETA`.
    const XI_ON_ZETA_SQRT: Self::BaseField;
```

Reply:
It seems to be `t^3` in the notations of [WB19, Section 4.2]. I don't understand what means "An arbitrary root of unity in the base field".

Reply:
From the file `ec/src/hasing/curve_maps/swu/mod.rs` line 149, I guess `ZETA` is a fixed fixed non-square ?

Reply:
a primitive root of unity I think is any generator of the multiplicative group and is a non-sequare. I think any non-square will do here. But this is coming from @daira 's code and I think it is chosen because there was already a constant time function which was using a pre defined root of unity to carry out this operation: https://github.com/arkworks-rs/algebra/pull/343/files#diff-b13fb8649e8438bfe74b4b6f00cf6e78debf4df9bc5fae8d56775e3ee15966adR128

We could use any other non-square including the one suggested in the paper  but I stuck to hir implementation so I'm able to run it on also pasta and make sure I get the same point as the one from @daira so I know the implementation is correct.